proto/v1beta1/intoto.proto

```diff
@@ -0,0 +1,163 @@
+option java_multiple_files = true;
+option java_package = "io.grafeas.v1beta1.intoto";
+option objc_class_prefix = "GRA";
+
+// This file defines the proto that matches the schema of in-toto metadata.
+// in-toto is a framework to protect a software supply chain. You can read more
+// about in-toto at https://in-toto.io/ and you can find the current version of
+// the in-toto specification at
+// https://github.com/in-toto/docs/blob/master/in-toto-spec.md
+// There are two types of in-toto metadata - layouts and links. Layouts define
+// the structure of the software supply chain, while links are signed records of
+// performing a step in the software supply chain.
+
+// This contains the fields corresponding to the definition of a software supply
+// chain step in an in-toto layout. This information goes into a Grafeas note.
+message InToto {
+  // This field identifies the name of the step in the supply chain.
+  string step_name = 1;
+
+  // This field contains the public keys that can be used to verify the
+  // signatures on the step metadata.
+  repeated SigningKey signing_keys = 2;
+
+  // Defines an object to declare an in-toto artifact rule
+  message ArtifactRule {
+    repeated string artifact_rule = 1;
+  }
+
+  // The following fields contain in-toto artifact rules identifying the
+  // artifacts that enter this supply chain step, and exit the supply chain
+  // step, i.e. materials and products of the step.
+  repeated ArtifactRule expected_materials = 3;
+  repeated ArtifactRule expected_products = 4;
+
+  // This field contains the expected command used to perform the step.
+  repeated string expected_command = 5;
+
+  // This field contains a value that indicates the minimum number of keys that
+  // need to be used to sign the step's in-toto link.
+  int64 threshold = 6;
+}
+
+// This defines the format used to record keys used in the software supply
+// chain. An in-toto link is attested using one or more keys defined in the
+// in-toto layout.
+message SigningKey {
+  // key_id is an identifier for the signing key.
+  string key_id = 1;
+
+  // This field identifies the hash algorithms used in the particular key ID.
+  repeated string key_id_hash_algorithms = 2;
+
+  // This field identifies the specific signing method. Eg: "RSASSA-PSS",
+  // "ed25519", and "ecdsa".
+  string key_type = 3;
+
+  // This field contains the actual public key.
+  string public_key_value = 4;
+
+  // This field contains the overall scheme used for the creation of the signing
+  // key. Eg: "rsassa-pss-sha256".
+  string key_scheme = 5;
+}
+
+// This corresponds to a signed in-toto link - it is made up of one or more
+// signatures and the in-toto link itself. This is used for occurrences of a
+// Grafeas in-toto note.
+message Details {
+  repeated Signature signatures = 1;
+  Link link = 2 [json_name="signed"];
+}
+
+// A signature object consists of the KeyID used and the signature itself.
+message Signature {
```

Comment:
While the format matches the in-toto definition, I think it goes back to a couple of things:
1. in-toto key IDs are not RFC3986 compliant - they're just hashes.
2. Further, they're also not RFC6920 compliant - while this can probably be handled in the conversion layer, there's also the matter that it's still a proposal.

Reply:
Makes sense, thanks